sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -559,6 +561,17 @@ object RemoveRedundantAliases extends Rule[LogicalPlan] {
         })
         Join(newLeft, newRight, joinType, newCondition, hint)
 
+      case _: Union =>
```

Comment:
Looking at the code again, I think for Union, we can remove aliases as long as it does not introduce conflicting attr IDs. For example, `Union(Project(a#1, a#1 as a#2, ...), ...)`, we can't remove the alias. But for `Union(Project(a#1 as a#2, ...), ...)`, we can.

If this was when we first add this rule, I won't bother to have unnecessary alias in the first child of Union. But this is a bug fix, and I think don't lose any optimization we had before is safer.

Reply:
All right, I will fix it today.

Reply:
fixed in https://github.com/apache/spark/pull/37334/commits/6adbd44197cbe8d9760f677b765ec94fd8e8745b